indykite/resource_external_data_resolver.go

```diff
@@ -0,0 +1,229 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indykite
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/indykite/indykite-sdk-go/config"
+	configpb "github.com/indykite/indykite-sdk-go/gen/indykite/config/v1beta1"
+)
+
+const (
+	externalDataResolverURLKey              = "url"
+	externalDataResolverMethodKey           = "method"
+	externalDataResolverHeadersKey          = "headers"
+	externalDataResolverRequestTypeKey      = "request_type"
+	externalDataResolverRequestPayloadKey   = "request_payload"
+	externalDataResolverResponseTypeKey     = "response_type"
+	externalDataResolverResponseSelectorKey = "response_selector"
+)
+
+var (
+	externalDataResolverHeaderRegex = regexp.MustCompile(`^[a-zA-Z0-9-]+$`)
+)
+
+func resourceExternalDataResolver() *schema.Resource {
+	readContext := configReadContextFunc(resourceExternalDataResolverFlatten)
+
+	return &schema.Resource{
+		Description: "ExternalDataResolver is a configuration node that allows to fetch data from external sources",
+
+		CreateContext: configCreateContextFunc(resourceExternalDataResolverBuild, readContext),
+		ReadContext:   readContext,
+		UpdateContext: configUpdateContextFunc(resourceExternalDataResolverBuild, readContext),
+		DeleteContext: configDeleteContextFunc(),
+		Importer: &schema.ResourceImporter{
+			StateContext: basicStateImporter,
+		},
+
+		Timeouts: defaultTimeouts(),
+		Schema: map[string]*schema.Schema{
+			locationKey:   locationSchema(),
+			customerIDKey: setComputed(customerIDSchema()),
+			appSpaceIDKey: setComputed(appSpaceIDSchema()),
+
+			nameKey:        nameSchema(),
+			displayNameKey: displayNameSchema(),
+			descriptionKey: descriptionSchema(),
+			createTimeKey:  createTimeSchema(),
+			updateTimeKey:  updateTimeSchema(),
+
+			externalDataResolverURLKey: {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.All(
+					validation.StringIsNotEmpty,
+					validation.IsURLWithHTTPorHTTPS,
+				),
+				Description: "Full URL to endpoint that will be called",
+			},
+			externalDataResolverMethodKey: {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.All(
+					validation.StringIsNotEmpty,
+					validation.StringInSlice([]string{"GET", "POST", "PUT", "PATCH"}, false),
```

Comment:
question: we allow `POST`, `PUT` and `PATCH`?

Reply:
currently we don't really have the fetcher implemented so not sure what we will support. I will leave it like that and maybe come back to update it when backend is implemented

Reply:
the design doc doesn't address it specifically, but I thought the plan for phase 1 was to just support `GET`. @arxeiss @jakubtomany ?

Reply:
I am allowing GET only and if we need to add more we will make terraform PR's

Reply:
We should update design doc then. Supporting only GET is not enough as you cannot send payload with it.

So yes, we support all those mentioned.

Reply:
OK, so I checked https://www.notion.so/indykite/Data-Reference-external-data-Technical-Design-Phase-1-d81af130e76d4b85a251e36b8cf37ad9 and there is no mention about supported/unsupported HTTP Methods. cc @maaland. So I would expect a fix by @Septimus4

Reply:
thanks for clarifying